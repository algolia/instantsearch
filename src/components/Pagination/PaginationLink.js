/** @jsx h */

import { h } from 'preact';
import PropTypes from 'prop-types';

function PaginationLink({
  cssClasses,
  label,
  ariaLabel,
  url,
  isDisabled,
  handleClick,
  pageNumber,
}) {
  if (isDisabled) {
    return (
      <li className={cssClasses.item}>
        <span
          className={cssClasses.link}
          dangerouslySetInnerHTML={{
            __html: label,
          }}
        />
      </li>
    );
  }

  return (
    <li className={cssClasses.item}>
      <a
        className={cssClasses.link}
        aria-label={ariaLabel}
        href={url}
        onClick={event => handleClick(pageNumber, event)}
        dangerouslySetInnerHTML={{
          __html: label,
        }}
      />
    </li>
  );
}

PaginationLink.propTypes = {
  ariaLabel: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
  cssClasses: PropTypes.shape({
    item: PropTypes.string.isRequired,
    link: PropTypes.string.isRequired,
  }).isRequired,
  handleClick: PropTypes.func.isRequired,
  isDisabled: PropTypes.bool,
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  pageNumber: PropTypes.number,
  url: PropTypes.string,
};

export default PaginationLink;
