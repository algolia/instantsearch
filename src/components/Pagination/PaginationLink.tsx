/** @jsx h */

import { h } from 'preact';

export type PaginationLinkCSSClasses = {
  item: string;
  link: string;
};

export type PaginationLinkProps = {
  ariaLabel: string;
  cssClasses: PaginationLinkCSSClasses;
  handleClick(pageNumber: number, event: MouseEvent): void;
  isDisabled: boolean;
  label: string;
  pageNumber: number;
  url?: string;
};

const PaginationLink = ({
  cssClasses,
  label,
  ariaLabel,
  url,
  isDisabled,
  handleClick,
  pageNumber,
}: PaginationLinkProps) => {
  if (isDisabled) {
    return (
      <li className={cssClasses.item}>
        <span
          className={cssClasses.link}
          dangerouslySetInnerHTML={{
            __html: label,
          }}
        />
      </li>
    );
  }

  return (
    <li className={cssClasses.item}>
      <a
        className={cssClasses.link}
        aria-label={ariaLabel}
        href={url}
        onClick={event => handleClick(pageNumber, event)}
        dangerouslySetInnerHTML={{
          __html: label,
        }}
      />
    </li>
  );
};

export default PaginationLink;
