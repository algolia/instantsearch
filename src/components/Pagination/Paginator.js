import range from 'lodash/utility/range';

class Paginator {
  constructor(params) {
    this.currentPage = params.currentPage;
    this.total = params.total;
    this.padding = params.padding;
  }

  pages() {
    const {total, currentPage, padding} = this;

    const totalDisplayedPages = this.nbPagesDisplayed(padding, total);
    if (totalDisplayedPages === total) return range(0, total);

    const paddingLeft = this.calculatePaddingLeft(currentPage, padding, total, totalDisplayedPages);
    const paddingRight = totalDisplayedPages - paddingLeft;

    const first = currentPage - paddingLeft;
    const last = currentPage + paddingRight;

    return range(first, last);
  }

  nbPagesDisplayed(padding, total) {
    return Math.min(2 * padding + 1, total);
  }

  calculatePaddingLeft(current, padding, total, totalDisplayedPages) {
    if (current <= padding) {
      return current;
    }

    if (current >= total - padding) {
      return totalDisplayedPages - (total - current);
    }

    return padding;
  }

  isLastPage() {
    return this.currentPage === this.total - 1;
  }

  isFirstPage() {
    return this.currentPage === 0;
  }
}

export default Paginator;
