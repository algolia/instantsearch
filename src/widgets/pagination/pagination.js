import React from 'react';
import ReactDOM from 'react-dom';
import defaults from 'lodash/object/defaults';
import cx from 'classnames';
import {
  bemHelper,
  getContainerNode
} from '../../lib/utils.js';
import autoHideContainerHOC from '../../decorators/autoHideContainer.js';
import PaginationComponent from '../../components/Pagination/Pagination.js';

const defaultLabels = {
  previous: '‹',
  next: '›',
  first: '«',
  last: '»'
};
const bem = bemHelper('ais-pagination');

/**
 * Add a pagination menu to navigate through the results
 * @function pagination
 * @param  {string|DOMElement} options.container CSS Selector or DOMElement to insert the widget
 * @param  {Object} [options.labels] Text to display in the various links (prev, next, first, last)
 * @param  {string} [options.labels.previous] Label for the Previous link
 * @param  {string} [options.labels.next] Label for the Next link
 * @param  {string} [options.labels.first] Label for the First link
 * @param  {string} [options.labels.last] Label for the Last link
 * @param  {number} [options.maxPages] The max number of pages to browse
 * @param  {number} [options.padding=3] The number of pages to display on each side of the current page
 * @param  {string|DOMElement|boolean} [options.scrollTo='body'] Where to scroll after a click, set to `false` to disable
 * @param  {boolean} [options.showFirstLast=true] Define if the First and Last links should be displayed
 * @param  {boolean} [options.autoHideContainer=true] Hide the container when no results match
 * @param  {Object} [options.cssClasses] CSS classes to be added
 * @param  {string|string[]} [options.cssClasses.root] CSS classes added to the parent `<ul>`
 * @param  {string|string[]} [options.cssClasses.item] CSS classes added to each `<li>`
 * @param  {string|string[]} [options.cssClasses.link] CSS classes added to each link
 * @param  {string|string[]} [options.cssClasses.page] CSS classes added to page `<li>`
 * @param  {string|string[]} [options.cssClasses.previous] CSS classes added to the previous `<li>`
 * @param  {string|string[]} [options.cssClasses.next] CSS classes added to the next `<li>`
 * @param  {string|string[]} [options.cssClasses.first] CSS classes added to the first `<li>`
 * @param  {string|string[]} [options.cssClasses.last] CSS classes added to the last `<li>`
 * @param  {string|string[]} [options.cssClasses.active] CSS classes added to the active `<li>`
 * @param  {string|string[]} [options.cssClasses.disabled] CSS classes added to the disabled `<li>`
 * @return {Object}
 */
const usage = `Usage:
pagination({
  container,
  [ cssClasses.{root,item,page,previous,next,first,last,active,disabled}={} ],
  [ labels.{previous,next,first,last} ],
  [ maxPages ],
  [ padding=3 ],
  [ showFirstLast=true ],
  [ autoHideContainer=true ],
  [ scrollTo='body' ]
})`;
function pagination({
    container,
    cssClasses: userCssClasses = {},
    labels: userLabels = {},
    maxPages,
    padding = 3,
    showFirstLast = true,
    autoHideContainer = true,
    scrollTo: userScrollTo = 'body'
  } = {}) {
  let scrollTo = userScrollTo;

  if (!container) {
    throw new Error(usage);
  }

  if (scrollTo === true) {
    scrollTo = 'body';
  }

  const containerNode = getContainerNode(container);
  const scrollToNode = scrollTo !== false ? getContainerNode(scrollTo) : false;
  let Pagination = PaginationComponent;
  if (autoHideContainer === true) {
    Pagination = autoHideContainerHOC(Pagination);
  }

  let cssClasses = {
    root: cx(bem(null), userCssClasses.root),
    item: cx(bem('item'), userCssClasses.item),
    link: cx(bem('link'), userCssClasses.link),
    page: cx(bem('item', 'page'), userCssClasses.page),
    previous: cx(bem('item', 'previous'), userCssClasses.previous),
    next: cx(bem('item', 'next'), userCssClasses.next),
    first: cx(bem('item', 'first'), userCssClasses.first),
    last: cx(bem('item', 'last'), userCssClasses.last),
    active: cx(bem('item', 'active'), userCssClasses.active),
    disabled: cx(bem('item', 'disabled'), userCssClasses.disabled)
  };

  let labels = defaults(userLabels, defaultLabels);

  return {
    init({helper}) {
      this.setCurrentPage = page => {
        helper.setCurrentPage(page);
        if (scrollToNode !== false) {
          scrollToNode.scrollIntoView();
        }
        helper.search();
      };
    },

    getMaxPage(results) {
      if (maxPages !== undefined) {
        return Math.min(maxPages, results.nbPages);
      }
      return results.nbPages;
    },

    render({results, state, createURL}) {
      ReactDOM.render(
        <Pagination
          createURL={page => createURL(state.setPage(page))}
          cssClasses={cssClasses}
          currentPage={results.page}
          labels={labels}
          nbHits={results.nbHits}
          nbPages={this.getMaxPage(results)}
          padding={padding}
          setCurrentPage={this.setCurrentPage}
          shouldAutoHideContainer={results.nbHits === 0}
          showFirstLast={showFirstLast}
        />,
        containerNode
      );
    }
  };
}

export default pagination;
