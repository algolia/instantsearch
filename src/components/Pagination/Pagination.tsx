/** @jsx h */

import { Component, h } from 'preact';
import cx from 'classnames';

import PaginationLink from './PaginationLink';
import { isSpecialClick } from '../../lib/utils';
import {
  PaginationCSSClasses,
  PaginationTemplates,
} from '../../widgets/pagination/pagination';
import { ComponentCSSClasses } from '../../types';

export type PaginationComponentCSSClasses = ComponentCSSClasses<
  PaginationCSSClasses
>;

export type PaginationComponentTemplates = Required<PaginationTemplates>;

export type PaginationProps = {
  createURL(value: number): string;
  cssClasses: PaginationComponentCSSClasses;
  currentPage: number;
  templates: PaginationComponentTemplates;
  nbPages?: number;
  pages?: number[];
  isFirstPage: boolean;
  isLastPage: boolean;
  setCurrentPage(value: number): void;
  showFirst?: boolean;
  showLast?: boolean;
  showPrevious?: boolean;
  showNext?: boolean;
};

type PageLink = {
  label: string;
  ariaLabel: string;
  pageNumber: number;
  additionalClassName: string | null;
  isDisabled?: boolean;
  isSelected?: boolean;
  createURL(value: number): string;
};

const defaultProps = {
  currentPage: 0,
  nbPages: 0,
  pages: [],
};

class Pagination extends Component<PaginationProps> {
  public static defaultProps = defaultProps;

  private pageLink({
    label,
    ariaLabel,
    pageNumber,
    additionalClassName = null,
    isDisabled = false,
    isSelected = false,
    createURL,
  }: PageLink) {
    const cssClasses = {
      item: cx(this.props.cssClasses.item, additionalClassName),
      link: this.props.cssClasses.link,
    };

    if (isDisabled) {
      cssClasses.item = cx(cssClasses.item, this.props.cssClasses.disabledItem);
    } else if (isSelected) {
      cssClasses.item = cx(cssClasses.item, this.props.cssClasses.selectedItem);
    }

    const url = !isDisabled ? createURL(pageNumber) : '#';

    return (
      <PaginationLink
        ariaLabel={ariaLabel}
        cssClasses={cssClasses}
        handleClick={this.handleClick}
        isDisabled={isDisabled}
        key={label + pageNumber + ariaLabel}
        label={label}
        pageNumber={pageNumber}
        url={url}
      />
    );
  }

  public handleClick = (pageNumber: number, event: MouseEvent) => {
    if (isSpecialClick(event)) {
      // do not alter the default browser behavior
      // if one special key is down
      return;
    }
    event.preventDefault();
    this.props.setCurrentPage(pageNumber);
  };

  private previousPageLink = () => {
    return this.pageLink({
      ariaLabel: 'Previous',
      additionalClassName: this.props.cssClasses.previousPageItem,
      isDisabled: this.props.isFirstPage,
      label: this.props.templates.previous,
      pageNumber: this.props.currentPage - 1,
      createURL: this.props.createURL,
    });
  };

  private nextPageLink = () => {
    return this.pageLink({
      ariaLabel: 'Next',
      additionalClassName: this.props.cssClasses.nextPageItem,
      isDisabled: this.props.isLastPage,
      label: this.props.templates.next,
      pageNumber: this.props.currentPage + 1,
      createURL: this.props.createURL,
    });
  };

  private firstPageLink = () => {
    return this.pageLink({
      ariaLabel: 'First',
      additionalClassName: this.props.cssClasses.firstPageItem,
      isDisabled: this.props.isFirstPage,
      label: this.props.templates.first,
      pageNumber: 0,
      createURL: this.props.createURL,
    });
  };

  private lastPageLink = () => {
    return this.pageLink({
      ariaLabel: 'Last',
      additionalClassName: this.props.cssClasses.lastPageItem,
      isDisabled: this.props.isLastPage,
      label: this.props.templates.last,
      pageNumber: this.props.nbPages! - 1,
      createURL: this.props.createURL,
    });
  };

  private pages = () => {
    return this.props.pages!.map(pageNumber =>
      this.pageLink({
        ariaLabel: `${pageNumber + 1}`,
        additionalClassName: this.props.cssClasses.pageItem,
        isSelected: pageNumber === this.props.currentPage,
        label: `${pageNumber + 1}`,
        pageNumber,
        createURL: this.props.createURL,
      })
    );
  };

  public render() {
    return (
      <div
        className={cx(this.props.cssClasses.root, {
          [this.props.cssClasses.noRefinementRoot]: this.props.nbPages! <= 1,
        })}
      >
        <ul className={this.props.cssClasses.list}>
          {this.props.showFirst && this.firstPageLink()}
          {this.props.showPrevious && this.previousPageLink()}
          {this.pages()}
          {this.props.showNext && this.nextPageLink()}
          {this.props.showLast && this.lastPageLink()}
        </ul>
      </div>
    );
  }
}

export default Pagination;
