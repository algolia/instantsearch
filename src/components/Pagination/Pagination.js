import React from 'react';
import forEach from 'lodash/collection/forEach';
import defaultsDeep from 'lodash/object/defaultsDeep';
import {isSpecialClick} from '../../lib/utils.js';

import Paginator from './Paginator.js';
import PaginationLink from './PaginationLink.js';

import cx from 'classnames';

class Pagination extends React.Component {
  constructor(props) {
    super(defaultsDeep(props, Pagination.defaultProps));
    this.handleClick = this.handleClick.bind(this);
  }

  pageLink({
    label,
    ariaLabel,
    pageNumber,
    additionalClassName = null,
    isDisabled = false,
    isActive = false,
    createURL
  }) {
    let cssClasses = {
      item: cx(this.props.cssClasses.item, additionalClassName),
      link: cx(this.props.cssClasses.link)
    };
    if (isDisabled) {
      cssClasses.item = cx(cssClasses.item, this.props.cssClasses.disabled);
    } else if (isActive) {
      cssClasses.item = cx(cssClasses.item, this.props.cssClasses.active);
    }

    let url = createURL && !isDisabled ? createURL(pageNumber) : '#';

    return (
      <PaginationLink
        ariaLabel={ariaLabel}
        cssClasses={cssClasses}
        handleClick={this.handleClick}
        isDisabled={isDisabled}
        key={label + pageNumber}
        label={label}
        pageNumber={pageNumber}
        url={url}
      />
    );
  }

  previousPageLink(pager, createURL) {
    return this.pageLink({
      ariaLabel: 'Previous',
      additionalClassName: this.props.cssClasses.previous,
      isDisabled: pager.isFirstPage(),
      label: this.props.labels.previous,
      pageNumber: pager.currentPage - 1,
      createURL
    });
  }

  nextPageLink(pager, createURL) {
    return this.pageLink({
      ariaLabel: 'Next',
      additionalClassName: this.props.cssClasses.next,
      isDisabled: pager.isLastPage(),
      label: this.props.labels.next,
      pageNumber: pager.currentPage + 1,
      createURL
    });
  }

  firstPageLink(pager, createURL) {
    return this.pageLink({
      ariaLabel: 'First',
      additionalClassName: this.props.cssClasses.first,
      isDisabled: pager.isFirstPage(),
      label: this.props.labels.first,
      pageNumber: 0,
      createURL
    });
  }

  lastPageLink(pager, createURL) {
    return this.pageLink({
      ariaLabel: 'Last',
      additionalClassName: this.props.cssClasses.last,
      isDisabled: pager.isLastPage(),
      label: this.props.labels.last,
      pageNumber: pager.total - 1,
      createURL
    });
  }

  pages(pager, createURL) {
    const pages = [];

    forEach(pager.pages(), pageNumber => {
      const isActive = pageNumber === pager.currentPage;

      pages.push(this.pageLink({
        ariaLabel: pageNumber + 1,
        additionalClassName: this.props.cssClasses.page,
        isActive,
        label: pageNumber + 1,
        pageNumber,
        createURL
      }));
    });

    return pages;
  }

  handleClick(pageNumber, event) {
    if (isSpecialClick(event)) {
      // do not alter the default browser behavior
      // if one special key is down
      return;
    }
    event.preventDefault();
    this.props.setCurrentPage(pageNumber);
  }

  render() {
    const pager = new Paginator({
      currentPage: this.props.currentPage,
      total: this.props.nbPages,
      padding: this.props.padding
    });

    const createURL = this.props.createURL;

    return (
      <ul className={this.props.cssClasses.root}>
        {this.props.showFirstLast ? this.firstPageLink(pager, createURL) : null}
        {this.previousPageLink(pager, createURL)}
        {this.pages(pager, createURL)}
        {this.nextPageLink(pager, createURL)}
        {this.props.showFirstLast ? this.lastPageLink(pager, createURL) : null}
      </ul>
    );
  }
}

Pagination.propTypes = {
  createURL: React.PropTypes.func,
  cssClasses: React.PropTypes.shape({
    root: React.PropTypes.string,
    item: React.PropTypes.string,
    link: React.PropTypes.string,
    page: React.PropTypes.string,
    previous: React.PropTypes.string,
    next: React.PropTypes.string,
    first: React.PropTypes.string,
    last: React.PropTypes.string,
    active: React.PropTypes.string,
    disabled: React.PropTypes.string
  }),
  currentPage: React.PropTypes.number,
  labels: React.PropTypes.shape({
    first: React.PropTypes.string,
    last: React.PropTypes.string,
    next: React.PropTypes.string,
    previous: React.PropTypes.string
  }),
  nbHits: React.PropTypes.number,
  nbPages: React.PropTypes.number,
  padding: React.PropTypes.number,
  setCurrentPage: React.PropTypes.func.isRequired,
  showFirstLast: React.PropTypes.bool
};

Pagination.defaultProps = {
  nbHits: 0,
  currentPage: 0,
  nbPages: 0
};

export default Pagination;
