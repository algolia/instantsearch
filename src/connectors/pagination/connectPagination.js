import {
  checkRendering,
  createDocumentationMessageGenerator,
  noop,
} from '../../lib/utils';
import Paginator from './Paginator';

const withUsage = createDocumentationMessageGenerator({
  name: 'pagination',
  connector: true,
});

/**
 * @typedef {Object} CustomPaginationWidgetOptions
 * @property {number} [totalPages] The total number of pages to browse.
 * @property {number} [padding = 3] The padding of pages to show around the current page
 */

/**
 * @typedef {Object} PaginationRenderingOptions
 * @property {function(page): string} createURL Creates URLs for the next state, the number is the page to generate the URL for.
 * @property {number} currentRefinement The number of the page currently displayed.
 * @property {number} nbHits The number of hits computed for the last query (can be approximated).
 * @property {number} nbPages The number of pages for the result set.
 * @property {number[]} pages The actual pages relevant to the current situation and padding
 * @property {boolean} isFirstPage true if the current page is also the first page
 * @property {boolean} isLastPage true if the current page is also the last page
 * @property {function(page)} refine Sets the current page and trigger a search.
 * @property {Object} widgetParams All original `CustomPaginationWidgetOptions` forwarded to the `renderFn`.
 */

/**
 * **Pagination** connector provides the logic to build a widget that will let the user
 * choose the current page of the results.
 *
 * When using the pagination with Algolia, you should be aware that the engine won't provide you pages
 * beyond the 1000th hits by default. You can find more information on the [Algolia documentation](https://www.algolia.com/doc/guides/searching/pagination/#pagination-limitations).
 *
 * @type {Connector}
 * @param {function(PaginationRenderingOptions, boolean)} renderFn Rendering function for the custom **Pagination** widget.
 * @param {function} unmountFn Unmount function called when the widget is disposed.
 * @return {function(CustomPaginationWidgetOptions)} Re-usable widget factory for a custom **Pagination** widget.
 * @example
 * // custom `renderFn` to render the custom Pagination widget
 * function renderFn(PaginationRenderingOptions, isFirstRendering) {
 *   if (isFirstRendering) {
 *     PaginationRenderingOptions.widgetParams.containerNode.html('<ul></ul>');
 *   }
 *
 *   // remove event listeners before replacing markup
 *   PaginationRenderingOptions.widgetParams.containerNode
 *     .find('a[data-page]')
 *     .each(function() { $(this).off('click'); });
 *
 *   var pages = PaginationRenderingOptions.pages
 *     .map(function(page) {
 *       return '<li style="display: inline-block; margin-right: 10px;">' +
 *         '<a href="' + PaginationRenderingOptions.createURL(page) + '" data-page="' + page + '">' +
 *         (parseInt(page) + 1) + '</a></li>';
 *     });
 *
 *   PaginationRenderingOptions.widgetParams.containerNode
 *     .find('ul')
 *     .html(pages);
 *
 *   PaginationRenderingOptions.widgetParams.containerNode
 *     .find('a[data-page]')
 *     .each(function() {
 *       $(this).on('click', function(event) {
 *         event.preventDefault();
 *         PaginationRenderingOptions.refine($(this).data('page'));
 *       });
 *     });
 * }
 *
 * // connect `renderFn` to Pagination logic
 * var customPagination = instantsearch.connectors.connectPagination(renderFn);
 *
 * // mount widget on the page
 * search.addWidget(
 *   customPagination({
 *     containerNode: $('#custom-pagination-container'),
 *     totalPages: 20,
 *     padding: 4,
 *   })
 * );
 */
export default function connectPagination(renderFn, unmountFn = noop) {
  checkRendering(renderFn, withUsage());

  return (widgetParams = {}) => {
    const { totalPages, padding = 3 } = widgetParams;

    const pager = new Paginator({
      currentPage: 0,
      total: 0,
      padding,
    });

    return {
      init({ helper, createURL, instantSearchInstance }) {
        this.refine = page => {
          helper.setPage(page);
          helper.search();
        };

        this.createURL = state => page => createURL(state.setPage(page));

        renderFn(
          {
            createURL: this.createURL(helper.state),
            currentRefinement: helper.state.page || 0,
            nbHits: 0,
            nbPages: 0,
            pages: [],
            isFirstPage: true,
            isLastPage: true,
            refine: this.refine,
            widgetParams,
            instantSearchInstance,
          },
          true
        );
      },

      getMaxPage({ nbPages }) {
        return totalPages !== undefined
          ? Math.min(totalPages, nbPages)
          : nbPages;
      },

      render({ results, state, instantSearchInstance }) {
        const page = state.page || 0;
        const nbPages = this.getMaxPage(results);
        pager.currentPage = page;
        pager.total = nbPages;

        renderFn(
          {
            createURL: this.createURL(state),
            currentRefinement: page,
            refine: this.refine,
            nbHits: results.nbHits,
            nbPages,
            pages: pager.pages(),
            isFirstPage: pager.isFirstPage(),
            isLastPage: pager.isLastPage(),
            widgetParams,
            instantSearchInstance,
          },
          false
        );
      },

      dispose({ state }) {
        unmountFn();

        return state.setQueryParameter('page', undefined);
      },

      getWidgetState(uiState, { searchParameters }) {
        const page = searchParameters.page || 0;

        if (page === 0 || page + 1 === uiState.page) {
          return uiState;
        }

        return {
          ...uiState,
          page: page + 1,
        };
      },

      getWidgetSearchParameters(searchParameters, { uiState }) {
        const uiPage = uiState.page;

        if (uiPage) {
          return searchParameters.setQueryParameter('page', uiState.page - 1);
        }

        return searchParameters.setQueryParameter('page', undefined);
      },
    };
  };
}
