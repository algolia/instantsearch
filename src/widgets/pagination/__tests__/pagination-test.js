/* eslint-env mocha */

import React from 'react';
import expect from 'expect';
import sinon from 'sinon';
import jsdom from 'jsdom-global';

import expectJSX from 'expect-jsx';
expect.extend(expectJSX);

import pagination from '../pagination';
import Pagination from '../../../components/Pagination/Pagination';

describe('pagination()', () => {
  beforeEach(function() {
    this.jsdom = jsdom();
  });
  afterEach(function() {
    this.jsdom();
  });

  describe('pagination call', () => {
    it('throws an exception when no container', () => {
      expect(pagination.bind(null)).toThrow(/^Usage/);
    });
  });

  describe('pagination()', () => {
    let ReactDOM;
    let container;
    let widget;
    let results;
    let helper;
    let cssClasses;

    beforeEach(() => {
      ReactDOM = {render: sinon.spy()};
      pagination.__Rewire__('ReactDOM', ReactDOM);
      pagination.__Rewire__('autoHideContainerHOC', sinon.stub().returns(Pagination));

      container = document.createElement('div');
      cssClasses = {
        root: ['root', 'cx'],
        item: 'item',
        link: 'link',
        page: 'page',
        previous: 'previous',
        next: 'next',
        first: 'first',
        last: 'last',
        active: 'active',
        disabled: 'disabled'
      };
      widget = pagination({container, scrollTo: false, cssClasses});
      results = {hits: [{first: 'hit', second: 'hit'}], nbHits: 200, hitsPerPage: 10, nbPages: 20};
      helper = {
        setCurrentPage: sinon.spy(),
        search: sinon.spy()
      };
      widget.init({helper});
    });

    it('configures nothing', () => {
      expect(widget.getConfiguration).toEqual(undefined);
    });

    it('sets the page', () => {
      widget.setCurrentPage(helper, 42);
      expect(helper.setCurrentPage.calledOnce).toBe(true);
      expect(helper.search.calledOnce).toBe(true);
    });

    it('calls twice ReactDOM.render(<Pagination props />, container)', () => {
      widget.render({results, helper});
      widget.render({results, helper});

      expect(ReactDOM.render.calledTwice).toBe(true, 'ReactDOM.render called twice');
      expect(ReactDOM.render.firstCall.args[0]).toEqualJSX(<Pagination {...getProps()} />);
      expect(ReactDOM.render.firstCall.args[1]).toEqual(container);
      expect(ReactDOM.render.secondCall.args[0]).toEqualJSX(<Pagination {...getProps()} />);
      expect(ReactDOM.render.secondCall.args[1]).toEqual(container);
    });

    context('mocking getContainerNode', function() {
      let scrollIntoView;

      beforeEach(() => {
        scrollIntoView = sinon.spy();
        const getContainerNode = sinon.stub().returns({
          scrollIntoView: scrollIntoView
        });
        pagination.__Rewire__('getContainerNode', getContainerNode);
      });

      it('should not scroll', () => {
        widget = pagination({container, scrollTo: false});
        widget.init({helper});
        widget.setCurrentPage(helper, 2);
        expect(scrollIntoView.calledOnce).toBe(false, 'scrollIntoView never called');
      });

      it('should scroll to body', () => {
        widget = pagination({container});
        widget.init({helper});
        widget.setCurrentPage(helper, 2);
        expect(scrollIntoView.calledOnce).toBe(true, 'scrollIntoView called once');
      });

      afterEach(() => {
        pagination.__ResetDependency__('utils');
      });
    });

    afterEach(() => {
      pagination.__ResetDependency__('ReactDOM');
      pagination.__ResetDependency__('autoHideContainerHOC');
    });

    function getProps() {
      return {
        cssClasses: {
          root: 'ais-pagination root cx',
          item: 'ais-pagination--item item',
          link: 'ais-pagination--link link',
          page: 'ais-pagination--item__page page',
          previous: 'ais-pagination--item__previous previous',
          next: 'ais-pagination--item__next next',
          first: 'ais-pagination--item__first first',
          last: 'ais-pagination--item__last last',
          active: 'ais-pagination--item__active active',
          disabled: 'ais-pagination--item__disabled disabled'
        },
        currentPage: 0,
        shouldAutoHideContainer: false,
        labels: {first: '«', last: '»', next: '›', previous: '‹'},
        nbHits: results.nbHits,
        nbPages: results.nbPages,
        padding: 3,
        setCurrentPage: () => {},
        showFirstLast: true,
        createURL: () => '#'
      };
    }
  });

  describe('pagination MaxPage', () => {
    let ReactDOM;
    let container;
    let widget;
    let results;
    let cssClasses;
    let paginationOptions;

    beforeEach(() => {
      ReactDOM = {render: sinon.spy()};
      pagination.__Rewire__('ReactDOM', ReactDOM);
      pagination.__Rewire__('autoHideContainerHOC', sinon.stub().returns(Pagination));

      container = document.createElement('div');
      cssClasses = {
        root: 'root',
        item: 'item',
        link: 'link',
        page: 'page',
        previous: 'previous',
        next: 'next',
        first: 'first',
        last: 'last',
        active: 'active',
        disabled: 'disabled'
      };
      results = {hits: [{first: 'hit', second: 'hit'}], nbHits: 300, hitsPerPage: 10, nbPages: 30};
      paginationOptions = {container, scrollTo: false, cssClasses};
    });

    it('does to have any default', () => {
      widget = pagination(paginationOptions);
      expect(widget.getMaxPage(results)).toEqual(30);
    });

    it('does reduce the number of page if lower than nbPages', () => {
      paginationOptions.maxPages = 20;
      widget = pagination(paginationOptions);
      expect(widget.getMaxPage(results)).toEqual(20);
    });

    it('does not reduce the number of page if greater than nbPages', () => {
      paginationOptions.maxPages = 40;
      widget = pagination(paginationOptions);
      expect(widget.getMaxPage(results)).toEqual(30);
    });
  });
});


