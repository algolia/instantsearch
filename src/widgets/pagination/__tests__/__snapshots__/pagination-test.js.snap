// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pagination() calls twice ReactDOM.render(<Pagination props />, container) 1`] = `
<Pagination
  createURL={[Function]}
  cssClasses={
    Object {
      "disabledItem": "ais-Pagination-item--disabled",
      "firstPageItem": "ais-Pagination-item--firstPage",
      "item": "ais-Pagination-item item",
      "lastPageItem": "ais-Pagination-item--lastPage",
      "link": "ais-Pagination-link link",
      "list": "ais-Pagination-list",
      "nextPageItem": "ais-Pagination-item--nextPage",
      "noRefinementRoot": "ais-Pagination--noRefinement",
      "pageItem": "ais-Pagination-item--page",
      "previousPageItem": "ais-Pagination-item--previousPage",
      "root": "ais-Pagination root cx",
      "selectedItem": "ais-Pagination-item--selected",
    }
  }
  currentPage={0}
  isFirstPage={true}
  isLastPage={false}
  labels={
    Object {
      "first": "«",
      "last": "»",
      "next": "›",
      "previous": "‹",
    }
  }
  nbHits={200}
  nbPages={20}
  pages={
    Array [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
    ]
  }
  setCurrentPage={[Function]}
  showFirst={true}
  showLast={true}
  showNext={true}
  showPrevious={true}
/>
`;

exports[`pagination() calls twice ReactDOM.render(<Pagination props />, container) 2`] = `
<Pagination
  createURL={[Function]}
  cssClasses={
    Object {
      "disabledItem": "ais-Pagination-item--disabled",
      "firstPageItem": "ais-Pagination-item--firstPage",
      "item": "ais-Pagination-item item",
      "lastPageItem": "ais-Pagination-item--lastPage",
      "link": "ais-Pagination-link link",
      "list": "ais-Pagination-list",
      "nextPageItem": "ais-Pagination-item--nextPage",
      "noRefinementRoot": "ais-Pagination--noRefinement",
      "pageItem": "ais-Pagination-item--page",
      "previousPageItem": "ais-Pagination-item--previousPage",
      "root": "ais-Pagination root cx",
      "selectedItem": "ais-Pagination-item--selected",
    }
  }
  currentPage={0}
  isFirstPage={true}
  isLastPage={false}
  labels={
    Object {
      "first": "«",
      "last": "»",
      "next": "›",
      "previous": "‹",
    }
  }
  nbHits={200}
  nbPages={20}
  pages={
    Array [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
    ]
  }
  setCurrentPage={[Function]}
  showFirst={true}
  showLast={true}
  showNext={true}
  showPrevious={true}
/>
`;
